native/src/main/java/io/ballerina/stdlib/data/csvdata/csv/CsvParser.java

```diff
@@ -0,0 +1,1075 @@
+                        sm.columnIndex++;
+                        state = this;
+                        continue;
+                    } else if (!sm.insideComment && ch == sm.config.textEnclosure) {
+                        sm.prevState = this;
+                        state = HEADER_QUOTE_CHAR_STATE;
+                        break;
+                    } else if (!sm.insideComment && ch == sm.config.escapeChar) {
+                        sm.prevState = this;
+                        state = HEADER_ESCAPE_CHAR_STATE;
+                        break;
+                    } else if (sm.insideComment && sm.isNewLineOrEof(ch)) {
+                        sm.insideComment = false;
+                        handleEndOfTheHeader(sm);
+                        state = HEADER_END_STATE;
+                    } else if (!sm.insideComment && isEndOfTheHeaderRow(sm, ch)) {
+                        handleEndOfTheHeader(sm);
+                        state = HEADER_END_STATE;
+                    } else if (StateMachine.isWhitespace(ch, sm.config.lineTerminator)) {
+                        if (sm.isValueStart) {
+                            sm.append(ch);
+                        }
+                        state = this;
+                        continue;
+                    } else {
+                        if (!sm.insideComment) {
+                            sm.append(ch);
+                            sm.isValueStart = true;
+                        }
+                        state = this;
+                        continue;
+                    }
+                    break;
+                }
+                if (state == null) {
+                    state = this;
+                }
+                sm.index = i + 1;
+                return state;
+            }
+        }
+
+        private static void handleEndOfTheHeader(StateMachine sm) throws CsvParserException {
+            handleEndOfTheHeader(sm, true);
+        }
+
+        private static void handleEndOfTheHeader(StateMachine sm, boolean trim) throws CsvParserException {
+            sm.isValueStart = false;
+            if (!sm.peek().isBlank()) {
+                addHeader(sm, trim);
+            }
+            finalizeHeaders(sm);
+            sm.columnIndex = 0;
+            sm.lineNumber++;
+        }
+
+        private static int getHeaderStartRowWhenHeaderIsPresent(Object header) {
+            return ((Long) header).intValue();
+        }
+
+        private static void finalizeHeaders(StateMachine sm) throws CsvParserException {
+            if (sm.headers.size() == 0) {
+                throw DiagnosticLog.error(DiagnosticErrorCode.HEADER_CANNOT_BE_EMPTY);
+            }
+            Type expType = sm.expectedArrayElementType;
+            if (expType instanceof RecordType) {
+                validateRemainingRecordFields(sm);
+            } else if (expType instanceof ArrayType) {
+                validateExpectedArraySize(((ArrayType) expType).getSize(), sm.headers.size());
```

Comment:
Prefer pattern matching to casting
```suggestion
            } else if (expType instanceof ArrayType arrayType) {
                validateExpectedArraySize(arrayType.getSize(), sm.headers.size());
```